File: api/app/domains/ledger/ledger_calc.py
"""Double-entry accounting core — pure, exact (integer paise), deterministic.

Implements the balance check, trial balance, P&L, balance sheet and depreciation
(SLM / WDV, Companies Act Schedule II). Normal balances: asset/expense are debit-natured;
liability/equity/income are credit-natured.
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal
from typing import Any

DEBIT_NATURED = ("asset", "expense")
CREDIT_NATURED = ("liability", "equity", "income")


def _round_rupee(paise: Decimal) -> int:
    return int((paise / 100).quantize(Decimal("1"), rounding=ROUND_HALF_UP)) * 100


def is_balanced(lines: list[dict]) -> bool:
    """A journal entry is valid only when total debits equal total credits."""
    return sum(int(ln.get("debit", 0)) for ln in lines) == sum(
        int(ln.get("credit", 0)) for ln in lines
    )


def trial_balance(lines: list[dict]) -> dict[str, Any]:
    """Totals across all posted lines. ``diff`` must be 0 for the books to tie out."""
    total_debit = sum(int(ln.get("debit", 0)) for ln in lines)
    total_credit = sum(int(ln.get("credit", 0)) for ln in lines)
    return {
        "total_debit": total_debit,
        "total_credit": total_credit,
        "diff": total_debit - total_credit,
        "balanced": total_debit == total_credit,
    }


def _net_by_nature(rows: list[dict], account_type: str) -> int:
    """Net movement for an account type. Debit-natured types return debit−credit;
    credit-natured return credit−debit."""
    total = 0
    for r in rows:
        if r["account_type"] != account_type:
            continue
        debit, credit = int(r.get("debit", 0)), int(r.get("credit", 0))
        total += (debit - credit) if account_type in DEBIT_NATURED else (credit - debit)
    return total


def profit_and_loss(rows: list[dict]) -> dict[str, int]:
    """rows: {account_type, debit, credit}. Returns income, expense, net_profit (paise)."""
    income = _net_by_nature(rows, "income")
    expense = _net_by_nature(rows, "expense")
    return {"income": income, "expense": expense, "net_profit": income - expense}


def balance_sheet(rows: list[dict]) -> dict[str, Any]:
    """Assets vs liabilities + equity + retained profit. ``balanced`` ⇔ the accounting
    equation holds."""
    assets = _net_by_nature(rows, "asset")
    liabilities = _net_by_nature(rows, "liability")
    equity = _net_by_nature(rows, "equity")
    net_profit = profit_and_loss(rows)["net_profit"]
    return {
        "assets": assets,
        "liabilities": liabilities,
        "equity": equity,
        "retained_profit": net_profit,
        "balanced": assets == liabilities + equity + net_profit,
    }


def slm_annual(cost: int, salvage: int, useful_life_years: int) -> int:
    """Straight-line depreciation per year = (cost − salvage) / life."""
    if useful_life_years <= 0:
        return 0
    return _round_rupee(Decimal(int(cost) - int(salvage)) / useful_life_years)


def wdv_annual(opening_wdv: int, rate_pct: float) -> int:
    """Written-down-value depreciation for one year = opening WDV × rate."""
    return _round_rupee(Decimal(int(opening_wdv)) * Decimal(str(rate_pct)) / 100)


def bank_reconciliation(
    book_balance: int,
    bank_statement_balance: int,
    *,
    deposits_in_transit: int = 0,
    unpresented_cheques: int = 0,
) -> dict[str, Any]:
    """Reconcile the cash-book balance to the bank statement. Adjusted bank balance =
    statement + deposits in transit − unpresented cheques; reconciled when it ties to the
    book balance (exact paise)."""
    adjusted = int(bank_statement_balance) + int(deposits_in_transit) - int(unpresented_cheques)
    difference = int(book_balance) - adjusted
    return {
        "book_balance": int(book_balance),
        "adjusted_bank_balance": adjusted,
        "difference": difference,
        "reconciled": difference == 0,
    }


# ── auto-posting: balanced journal-line builders for cross-module source events ──────
# Each returns lines for ``post_journal_entry`` and is internally balanced by construction.


def payroll_journal(
    *,
    salary_expense_account: int,
    bank_account: int,
    statutory_payable_account: int,
    gross: int,
    net: int,
    statutory: int,
) -> list[dict]:
    """Dr salary expense (gross); Cr bank (net pay); Cr statutory payable (PF/ESI/TDS withheld).
    Requires net + statutory == gross."""
    if int(net) + int(statutory) != int(gross):
        raise ValueError("payroll journal: net + statutory must equal gross")
    return [
        {
            "account_id": salary_expense_account,
            "debit": int(gross),
            "credit": 0,
            "description": "Payroll — gross salary",
        },
        {
            "account_id": bank_account,
            "debit": 0,
            "credit": int(net),
            "description": "Payroll — net pay",
        },
        {
            "account_id": statutory_payable_account,
            "debit": 0,
            "credit": int(statutory),
            "description": "Payroll — statutory payable",
        },
    ]


def sales_journal(
    *,
    receivable_account: int,
    sales_account: int,
    gst_output_account: int,
    taxable: int,
    tax: int,
) -> list[dict]:
    """Dr receivable (taxable + tax); Cr sales (taxable); Cr GST output (tax)."""
    return [
        {
            "account_id": receivable_account,
            "debit": int(taxable) + int(tax),
            "credit": 0,
            "description": "Invoice — receivable",
        },
        {
            "account_id": sales_account,
            "debit": 0,
            "credit": int(taxable),
            "description": "Invoice — sales",
        },
        {
            "account_id": gst_output_account,
            "debit": 0,
            "credit": int(tax),
            "description": "Invoice — GST output",
        },
    ]


def gst_payment_journal(*, gst_payable_account: int, bank_account: int, amount: int) -> list[dict]:
    """Dr GST payable; Cr bank — settling the net GST liability."""
    return [
        {
            "account_id": gst_payable_account,
            "debit": int(amount),
            "credit": 0,
            "description": "GST payment",
        },
        {
            "account_id": bank_account,
            "debit": 0,
            "credit": int(amount),
            "description": "GST payment — bank",
        },
    ]

File: api-nest/src/domains/ledger/ledger.calc.ts
/**
 * Double-entry accounting core — pure, exact (integer paise), deterministic.
 * Faithful port of api/app/domains/ledger/ledger_calc.py.
 *
 * Money is integer paise throughout (safe in JS numbers, well under 2^53).
 * Rupee rounding replicates Python's Decimal ROUND_HALF_UP (half away from zero);
 * all division is done in integer arithmetic to avoid float drift.
 * Normal balances: asset/expense are debit-natured; liability/equity/income are credit-natured.
 */

export const DEBIT_NATURED = ['asset', 'expense'] as const;
export const CREDIT_NATURED = ['liability', 'equity', 'income'] as const;

export type Line = { debit?: number; credit?: number };
export type TypedRow = { account_type: string; debit?: number; credit?: number };
export type JournalLineOut = {
  account_id: number;
  debit: number;
  credit: number;
  description: string;
};

/** Integer half-away-from-zero of n/d (d>0), matching Decimal ROUND_HALF_UP on rupees. */
function divRoundHalfUp(n: number, d: number): number {
  const sign = n < 0 ? -1 : 1;
  const a = Math.abs(n);
  return sign * Math.floor((2 * a + d) / (2 * d));
}

export function isBalanced(lines: Line[]): boolean {
  const dr = lines.reduce((s, ln) => s + ~~(ln.debit ?? 0), 0);
  const cr = lines.reduce((s, ln) => s + ~~(ln.credit ?? 0), 0);
  return dr === cr;
}

export function trialBalance(lines: Line[]) {
  const totalDebit = lines.reduce((s, ln) => s + ~~(ln.debit ?? 0), 0);
  const totalCredit = lines.reduce((s, ln) => s + ~~(ln.credit ?? 0), 0);
  return {
    total_debit: totalDebit,
    total_credit: totalCredit,
    diff: totalDebit - totalCredit,
    balanced: totalDebit === totalCredit,
  };
}

function netByNature(rows: TypedRow[], accountType: string): number {
  const debitNatured = (DEBIT_NATURED as readonly string[]).includes(accountType);
  let total = 0;
  for (const r of rows) {
    if (r.account_type !== accountType) continue;
    const debit = ~~(r.debit ?? 0);
    const credit = ~~(r.credit ?? 0);
    total += debitNatured ? debit - credit : credit - debit;
  }
  return total;
}

export function profitAndLoss(rows: TypedRow[]) {
  const income = netByNature(rows, 'income');
  const expense = netByNature(rows, 'expense');
  return { income, expense, net_profit: income - expense };
}

export function balanceSheet(rows: TypedRow[]) {
  const assets = netByNature(rows, 'asset');
  const liabilities = netByNature(rows, 'liability');
  const equity = netByNature(rows, 'equity');
  const netProfit = profitAndLoss(rows).net_profit;
  return {
    assets,
    liabilities,
    equity,
    retained_profit: netProfit,
    balanced: assets === liabilities + equity + netProfit,
  };
}

/** Straight-line depreciation per year = (cost − salvage) / life, rounded to whole rupee. */
export function slmAnnual(cost: number, salvage: number, usefulLifeYears: number): number {
  if (usefulLifeYears <= 0) return 0;
  const rupees = divRoundHalfUp(~~cost - ~~salvage, usefulLifeYears * 100);
  return rupees * 100;
}

/**
 * WDV depreciation for one year = opening WDV × rate%, rounded to whole rupee.
 * rate_pct is a JS number; String(rate_pct) mirrors Python str()→Decimal(str), so we
 * parse its decimal digits to keep integer arithmetic exact (no float 13.91 drift).
 */
export function wdvAnnual(openingWdv: number, ratePct: number): number {
  const s = String(ratePct);
  const [intPart, fracPart = ''] = s.replace('-', '').split('.');
  const k = fracPart.length;
  const numRate = parseInt((intPart + fracPart) || '0', 10) * (s.startsWith('-') ? -1 : 1);
  // paise = openingWdv * ratePct / 100; rupees = paise / 100 = openingWdv*numRate / (10^k * 10000)
  const denom = Math.pow(10, k) * 10000;
  const rupees = divRoundHalfUp(~~openingWdv * numRate, denom);
  return rupees * 100;
}

export function bankReconciliation(
  bookBalance: number,
  bankStatementBalance: number,
  opts: { deposits_in_transit?: number; unpresented_cheques?: number } = {},
) {
  const deposits = ~~(opts.deposits_in_transit ?? 0);
  const unpresented = ~~(opts.unpresented_cheques ?? 0);
  const adjusted = ~~bankStatementBalance + deposits - unpresented;
  const difference = ~~bookBalance - adjusted;
  return {
    book_balance: ~~bookBalance,
    adjusted_bank_balance: adjusted,
    difference,
    reconciled: difference === 0,
  };
}

// ── auto-posting: balanced journal-line builders for cross-module source events ──────

export function payrollJournal(args: {
  salary_expense_account: number;
  bank_account: number;
  statutory_payable_account: number;
  gross: number;
  net: number;
  statutory: number;
}): JournalLineOut[] {
  if (~~args.net + ~~args.statutory !== ~~args.gross) {
    throw new Error('payroll journal: net + statutory must equal gross');
  }
  return [
    { account_id: args.salary_expense_account, debit: ~~args.gross, credit: 0, description: 'Payroll — gross salary' },
    { account_id: args.bank_account, debit: 0, credit: ~~args.net, description: 'Payroll — net pay' },
    { account_id: args.statutory_payable_account, debit: 0, credit: ~~args.statutory, description: 'Payroll — statutory payable' },
  ];
}

export function salesJournal(args: {
  receivable_account: number;
  sales_account: number;
  gst_output_account: number;
  taxable: number;
  tax: number;
}): JournalLineOut[] {
  return [
    { account_id: args.receivable_account, debit: ~~args.taxable + ~~args.tax, credit: 0, description: 'Invoice — receivable' },
    { account_id: args.sales_account, debit: 0, credit: ~~args.taxable, description: 'Invoice — sales' },
    { account_id: args.gst_output_account, debit: 0, credit: ~~args.tax, description: 'Invoice — GST output' },
  ];
}

export function gstPaymentJournal(args: {
  gst_payable_account: number;
  bank_account: number;
  amount: number;
}): JournalLineOut[] {
  return [
    { account_id: args.gst_payable_account, debit: ~~args.amount, credit: 0, description: 'GST payment' },
    { account_id: args.bank_account, debit: 0, credit: ~~args.amount, description: 'GST payment — bank' },
  ];
}
